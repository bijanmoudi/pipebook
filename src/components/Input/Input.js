import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import './index.scss';

const Input = React.forwardRef(
	(
		{
			value,
			className,
			autoFocus,
			hasReset,
			onChange,
			onSubmit,
			onValidate,
			onStopTyping,
			stopTypingThreshold,
			validateOnStart,
			containerClassName,
			...restProps
		},
		ref
	) => {
		const [text, setText] = useState(value);
		const [errors, setError] = useState([]);
		const [typingTimeout, setTypingTimeout] = useState(null);
		const handleChange = val => {
			setText(val);
			onChange(val, ref.current);
			typingTimeout && clearTimeout(typingTimeout);
			setTypingTimeout(
				setTimeout(() => onStopTyping(val, ref.current), stopTypingThreshold)
			);
		};
		const handleValidation = input => {
			let errors = [];
			const isValid = input.checkValidity();
			if (!isValid) {
				errors.push(input.validationMessage);
			}
			setError(errors);
			onValidate(errors);
		};
		const handleSubmit = e => {
			e.key === 'Enter' &&
				document.activeElement === ref.current &&
				errors.length === 0 &&
				onSubmit(ref.current);
		};
		const handleReset = () => {
			handleChange('');
			ref.current.focus();
		};
		useEffect(() => {
			setText(value);
		}, [value]);
		useEffect(() => {
			ref.current && handleValidation(ref.current);
			// eslint-disable-next-line
		}, [text]);
		useEffect(() => {
			validateOnStart && handleValidation(ref.current);
			autoFocus && ref.current && ref.current.focus();
			// eslint-disable-next-line
		}, [ref.current]);
		useEffect(() => {
			window.addEventListener('keyup', handleSubmit);
			return () => {
				window.removeEventListener('keyup', handleSubmit);
			};
		});
		return (
			<div
				className={classnames('input', {
					[containerClassName]: containerClassName
				})}
			>
				<input
					ref={ref}
					value={text}
					onChange={e => handleChange(e.target.value)}
					className={classnames('input__element', {
						[className]: className
					})}
					{...restProps}
				/>
				{hasReset && text.length ? (
					<div className="input__reset-container">
						<button
							className="input__reset"
							title="Reset"
							onClick={handleReset}
						>
							<span className="visuallyhidden">Reset</span>
						</button>
					</div>
				) : null}
				{errors.length ? (
					<div className="input__tip" role="alert">
						<ul>
							{errors.map((error, key) => (
								<li key={`__error${key}`} className="input__tip-item">
									{error}
								</li>
							))}
						</ul>
					</div>
				) : null}
			</div>
		);
	}
);

Input.propTypes = {
	value: PropTypes.string,
	spellCheck: PropTypes.oneOf(['true', 'false', true, false]),
	autoCapitalize: PropTypes.oneOf(['on', 'off']),
	autoCorrect: PropTypes.oneOf(['on', 'off']),
	autoComplete: PropTypes.oneOf(['on', 'off']),
	autoFocus: PropTypes.bool,
	hasReset: PropTypes.bool,
	onChange: PropTypes.func,
	onSubmit: PropTypes.func,
	onValidate: PropTypes.func,
	onStopTyping: PropTypes.func,
	stopTypingThreshold: PropTypes.number,
	validateOnStart: PropTypes.bool,
	containerClassName: PropTypes.oneOfType([() => null, PropTypes.string])
};

Input.defaultProps = {
	value: '',
	spellCheck: 'false',
	autoCapitalize: 'off',
	autoCorrect: 'off',
	autoComplete: 'off',
	autoFocus: false,
	hasReset: true,
	onChange: () => {},
	onSubmit: () => {},
	onValidate: () => {},
	onStopTyping: () => {},
	stopTypingThreshold: 500,
	validateOnStart: false,
	containerClassName: null
};

export default Input;
