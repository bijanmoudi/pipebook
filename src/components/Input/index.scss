/*====================================
=            Input Styles            =
====================================*/

@import '../../assets/styles/constants/layout';
@import '../../assets/styles/helpers/colors';
@import '../../assets/styles/helpers/units';

.input {
	position: relative;
	&__element {
		width: 100%;
		height: em(39, 14);
		padding: em(8, 14) em(15, 14);
		font-size: em(14);
		background-color: get-color('gray-light-very');
		border: em(2, 14) solid get-color('gray-light-very');
		border-radius: rem($border-radius);
		line-height: 1;
		&::placeholder {
			line-height: normal;
		}
		&:hover,
		&:focus {
			background-color: get-color('gray-light-ultra');
		}
		&:focus {
			border-color: get-color('gray-light');
			&::placeholder {
				opacity: 0.5;
			}
		}
		&:focus ~ .input__tip {
			opacity: 1;
			visibility: visible;
		}
	}
	&__reset {
		&-container {
			z-index: 1;
			position: absolute;
			top: 0;
			bottom: 0;
			right: 0;
			width: rem(40);
		}
		position: absolute;
		top: calc(50% - #{em(11, 16)});
		right: em(8, 16);
		display: flex;
		align-items: center;
		justify-content: center;
		width: em(22, 16);
		height: em(22, 16);
		border-radius: 100%;
		border: 1px solid get-color('gray-light');
		background-color: get-color('white');
		color: get-color('gray-dark');
		line-height: em(19);
		font-size: em(16);
		font-weight: $semi-bold-font-weight;
		&:before {
			content: 'Ã—';
		}
		&:hover,
		&:focus {
			border-color: get-color('gray-dark');
		}
		&:focus {
			box-shadow: 0 0 0 rem(1);
		}
	}
	&__tip {
		position: absolute;
		top: 100%;
		left: 0;
		max-width: 100%;
		margin-top: em(7, 13);
		padding: em(7, 13) em(10, 13);
		border-radius: rem($border-radius);
		background-color: get-color('orange');
		font-weight: 600;
		font-size: em(13);
		line-height: 1;
		opacity: 0;
		visibility: hidden;
		transition: 250ms ease opacity, 250ms ease visibility;
		&:before {
			content: '';
			position: absolute;
			bottom: 100%;
			left: em(15, 13);
			width: 0;
			height: 0;
			border-style: solid;
			border-width: 0 em(5, 13) em(5, 13) em(5, 13);
			border-color: transparent transparent #{get-color('orange')} transparent;
			border-style: inset;
			-webkit-transform: rotate(360deg);
		}
	}
	&__tip-item {
		& + & {
			margin-top: 1em;
		}
	}
}
