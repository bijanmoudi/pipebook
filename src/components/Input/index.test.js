import React from 'react';
import renderer from 'react-test-renderer';
import Input from './Input';

const props = {
	value: 'Dulce',
	hasReset: false,
	containerClassName: 'search__input',
	ref: React.createRef()
};

describe('Testing "Input" component', () => {
	it('Rendering without default value but with custom container class', () => {
		const tree = renderer
			.create(
				<Input
					ref={props.ref}
					containerClassName={props.containerClassName}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering with default value and without reset button', () => {
		const tree = renderer
			.create(
				<Input
					ref={props.ref}
					value={props.value}
					hasReset={props.hasReset}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
