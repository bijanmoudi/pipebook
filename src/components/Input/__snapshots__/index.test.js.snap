// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing "Input" component Rendering with default value and without reset button 1`] = `
<div
  className="input"
>
  <input
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="input__element"
    onChange={[Function]}
    spellCheck="false"
    value="Dulce"
  />
</div>
`;

exports[`Testing "Input" component Rendering without default value but with custom container class 1`] = `
<div
  className="input search__input"
>
  <input
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="input__element"
    onChange={[Function]}
    spellCheck="false"
    value=""
  />
</div>
`;
